### ISSUE TITLE
[salt-cloud] openstack driver tries to assign new floating ip when some are available in pool

### ISSUE THREAD

**Issue by arthurlogilab** [saltstack/salt#31762]
### Description of Issue/Question

When creating VMs using salt-cloud against an icehouse openstack using the openstack driver, I get a "no more floating ips", it seems to be trying to allocate new floating IPs when some are available to assign.
### Steps to Reproduce Issue

Traceback : 

```
[ERROR   ] There was a profile error: 404 Not Found No more floating ips in pool [u'ext-net'].
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/cloud/cli.py", line 277, in run
    self.config.get('names')
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1418, in run_profile
    ret[name] = self.create(vm_)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1262, in create
    output = self.clouds[func](vm_)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/openstack.py", line 618, in create
    data, vm_ = request_instance(vm_)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/openstack.py", line 489, in request_instance
    floating = _assign_floating_ips(vm_, conn, kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/openstack.py", line 819, in _assign_floating_ips
    for idx in [pool.create_floating_ip()]:
  File "/usr/lib/python2.7/dist-packages/libcloud/compute/drivers/openstack.py", line 2336, in create_floating_ip
    data={'pool': self.name})
  File "/usr/lib/python2.7/dist-packages/libcloud/common/openstack.py", line 200, in request
    raw=raw)
  File "/usr/lib/python2.7/dist-packages/libcloud/common/base.py", line 736, in request
    response = responseCls(**kwargs)
  File "/usr/lib/python2.7/dist-packages/libcloud/compute/drivers/openstack.py", line 956, in __init__
    super(OpenStack_1_1_Response, self).__init__(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/libcloud/common/base.py", line 117, in __init__
    raise Exception(self.parse_error())
Exception: 404 Not Found No more floating ips in pool [u'ext-net'].
```
### Versions Report

```
$ salt --versions-report
Salt Version:
           Salt: 2015.8.7

Dependency Versions:
         Jinja2: 2.8
       M2Crypto: 0.21.1
           Mako: 1.0.2
         PyYAML: 3.11
          PyZMQ: 14.4.1
         Python: 2.7.10 (default, Oct 14 2015, 16:09:02)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 1.1.2
       cherrypy: 3.5.0
       dateutil: 2.2
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.3.4
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.9.0
        timelib: Not Installed

System Versions:
           dist: Ubuntu 15.10 wily
        machine: x86_64
        release: 4.2.0-30-generic
         system: Ubuntu 15.10 wily
```

**Comment by jfindlay** [comment]
@arthurlogilab, thanks for reporting.  What version of libcloud are you using (should appear in `salt-cloud --versions-report`)?

**Comment by arthurlogilab** [comment]
```$ salt-cloud --versions-report
Salt Version:
            Salt: 2015.8.7
 
Dependency Versions:
 Apache Libcloud: 0.17.0
          Jinja2: 2.8
        M2Crypto: 0.21.1
            Mako: 1.0.2
          PyYAML: 3.11
           PyZMQ: 14.4.1
          Python: 2.7.10 (default, Oct 14 2015, 16:09:02)
            RAET: Not Installed
         Tornado: 4.2.1
             ZMQ: 4.0.5
            cffi: 1.1.2
        cherrypy: 3.5.0
        dateutil: 2.2
           gitdb: 0.6.4
       gitpython: 1.0.1
           ioflo: Not Installed
         libgit2: Not Installed
         libnacl: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.4.2
    mysql-python: 1.3.4
       pycparser: 2.14
        pycrypto: 2.6.1
          pygit2: Not Installed
    python-gnupg: Not Installed
           smmap: 0.9.0
         timelib: Not Installed
 
System Versions:
            dist: Ubuntu 15.10 wily
         machine: x86_64
         release: 4.2.0-30-generic
          system: Ubuntu 15.10 wily
```

**Comment by jfindlay** [comment]
@arthurlogilab, thanks for the extra info.  You may have better fortune with the nova driver.

**Comment by seacrib** [comment]
This is caused by the attempted fix of #12225 to get parallel mode to work. I'd prefer that the commit be reverted. I think it is better being forced to do serial creation, rather than not being able to do any creation when there are no more IPs in a floating IP pool.

**Comment by jfindlay** [comment]
Ping @rallytime.

**Comment by rallytime** [comment]
This is probably something that @gtmanfred should look into and likely needs a fix separate from just reverting the changes made in the referenced #12225 above. That code just helps avoid a race condition and makes the code more DRY and has been there for a while.

**Comment by toanju** [comment]
just fell over this old issue which should be fixed fixed in #38965 or #34280, likely nova driver is the better choice though
